example/multistep/multistep.py

```diff
@@ -0,0 +1,92 @@
+"""
+Downloads the movielens dataset, ETLs it into Parquet, trains an
+ALS model, and uses the ALS model to train a Keras neural network.
+
+See README.rst for more details.
+"""
+
+import click
+import os
+import tempfile
+
+import mlflow
+from mlflow.entities import RunStatus, Run
+from mlflow.utils.logging_utils import eprint
+import six
+
+from mlflow.tracking.fluent import _get_experiment_id
+
+
+@click.group()
+def cli():
+    pass
+
+
+def _get_params(run):
+    """Converts [mlflow.entities.Param] to a dictionary of {k: v}."""
+    return {param.key: param.value for param in run.data.params}
+
+
+def _already_ran(entry_point_name, parameters, experiment_id=None):
+    """Best-effort detection of if a run with the given entrypoint name,
+    parameters, and experiment id already ran. The run must have completed
+    successfully and have at least the parameters provided.
+    """
+    experiment_id = experiment_id if experiment_id is not None else _get_experiment_id()
+    service = mlflow.tracking.get_service()
+    all_run_infos = reversed(service.list_run_infos(experiment_id))
+    for run_info in all_run_infos:
+        if run_info.entry_point_name != entry_point_name:
+            continue
+
+        full_run = service.get_run(run_info.run_uuid)
+        run_params = _get_params(full_run)
+        match_failed = False
+        for param_key, param_value in six.iteritems(parameters):
+            run_value = run_params.get(param_key)
+            if run_value != param_value:
+                match_failed = True
+                break
+        if match_failed:
+            continue
+
+        if run_info.status != RunStatus.FINISHED:
+            eprint(("Run matched, but is not FINISHED, so skipping "
+                    "(run_id=%s, status=%s)") % (run_info.run_uuid, run_info.status))
+            continue
+        return service.get_run(run_info.run_uuid)
+    return None
+
+
+# TODO(aaron): This is not great because it doesn't account for:
+# - changes in code
+# - changes in dependant steps
```

Comment:
Should we pass in the code version here and search for that too? I guess we'd need a way to figure out our current Git version, but we could maybe query the ActiveRun in MLflow for it.

Reply:
BTW I think the changes in dependent steps will automatically be covered to some extent if the parameters to these steps are run IDs. Then we might sometimes rerun more stuff than needed, but we would reuse a chain of steps if their params didn't change.

Reply:
Currently, we're explicitly not using run ids as parameters to allow running things using their lower-level constructs (i.e., a path to S3). We should revisit that.